Added Operational Insights Data Plane cmdlet

csuich2 opened the pull request:
This PR adds a cmdlet for the Operational Insights Data Plane query API.

The project that I'm adding this cmdlet to doesn't seem to be using the platyPS documentation system so I'd like to avoid introducing it if at all possible since I don't own the rest of the project.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [(see comment)] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

csuich2 commented:
The build isn't succeeding yet because the C# package hasn't been published. I'm looking in to that but thought I'd go ahead and open this PR to get some eyes on it.

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+        [Parameter(Mandatory = false, HelpMessage = "If specified, query statistics will be included in the response.")]
+        public SwitchParameter IncludeStatistics { get; set; }
+
+        [Parameter(Mandatory = false,
```
So when exactly is the parameter used? Does the user have a "DEMO_WORKSPACE"?

csuich2 replied:
So our API has a global workspace ID called "DEMO_WORKSPACE" that any customer can use as a sandbox/playground. The idea is to require the user to explicitly indicate they want to use the demo workspace here. I suppose that we could omit the switch and use the demo key anytime the workspace ID == "DEMO_WORKSPACE" if preferred.

maddieclayton replied:
Let's talk about this more in the cmdlet review - I think the idea that the key would be automatically used is better than having a parameter that might confuse users.

csuich2 replied:
Sure - sounds good.

markcowl replied:
Yes, this is preferred, please make this change

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+            HelpMessage = "When specified, uses the 'DEMO_KEY' API key for authentication. This is only valid when querying against the workspace 'DEMO_WORKSPACE'.")]
+        public SwitchParameter UseDemoKey { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "For internal use only.")]
```
Why do you have parameters for internal use only? Why would you want the user to be able to see these parameters and why do you need them?

csuich2 replied:
This wasn't desirable. I was continuing to search for a way to resolve this and settled on introducing another field in the AzureEnvironment. Originally this was to support testing against INT, but then I realized that having it in AzureEnvironment was the correct long term approach for national clouds.

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+        [Parameter(Mandatory = true, HelpMessage = "The query to execute.")]
+        public string Query { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "The timespan to bound the query by.")]
```
This seems rather unusual - is this necessary? If the user doesn't want to wait, they should create a powershell job I would think.

csuich2 replied:
The "Wait" here refers to how long the API should "wait" for the database to respond for a query. This is documented in the link that is included in the help message, but I'm open to renaming this parameter if you believe that will help.

maddieclayton replied:
Okay we can talk about it when we do a cmdlet review.

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/packages.config:
```diff
@@ -1,4 +1,9 @@
   <package id="Microsoft.Azure.Management.OperationalInsights" version="0.18.0-preview" targetFramework="net45" />
-</packages>
+  <package id="Microsoft.Azure.OperationalInsights" version="1.0.0-preview" targetFramework="net452" />
+  <package id="Microsoft.CSharp" version="4.3.0" targetFramework="net452" />
```
Remove the following four packages, they will be imported with common dependencies.

csuich2 replied:
Fixed.

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+﻿using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.OperationalInsights.Models;
```
Please add help for this cmdlet: https://github.com/Azure/azure-powershell/wiki/Azure-Powershell-Developer-Guide#adding-help-content

csuich2 replied:
Fixed.

maddieclayton commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace Microsoft.Azure.Commands.OperationalInsights.Query
+{
+    [Cmdlet("Invoke", "AzureRmOperationalInsightsQuery"), OutputType(typeof(PSQueryResponse))]
```
This should implement ShouldProcess: https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess

csuich2 replied:
That documentation seems to say ShouldProcess is only necessary if it changes the system, which our query API doesn't. Is there another reason we should support ShouldProcess?

maddieclayton replied:
Does this cmdlet create a query which is saved anywhere in the system? The use of the verb "invoke" and the function being called "create" led me to believe the query was created and saved somewhere, in which case this is a cmdlet making changes to the system.  Let me know if this is incorrect in which case SupportsShouldProcess is not necessary.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/Commands.OperationalInsights.Test.csproj:
```diff
@@ -61,6 +61,9 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.OperationalInsights.0.18.0-preview\lib\net45\Microsoft.Azure.Management.OperationalInsights.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="Microsoft.Azure.OperationalInsights, Version=0.18.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
The above package is the management plane while the new package is the separate management plane. The API specs & C# clients are separate since the data plane goes direct to our API instead of going through ARM.

maddieclayton replied:
Ah, of course.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Commands.OperationalInsights.csproj:
```diff
@@ -51,6 +51,18 @@
+    <Reference Include="Microsoft.Azure.OperationalInsights, Version=0.18.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.OperationalInsights.1.0.0-preview\lib\net452\Microsoft.Azure.OperationalInsights.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Ah, I see. They got added when I installed the new data plane package via NuGet.

maddieclayton commented:
@csuich2 Before we go any further, please fill out the cmdlet review - it will be easier to dive in deeper once we agree upon the overall design.

csuich2 commented on src/ResourceManager/Common/Commands.Common.Authentication.ResourceManager/Models/PSAzureEnvironment.cs:
```diff
@@ -261,7 +266,8 @@ public override bool Equals(object obj)
                        && TrafficManagerDnsSuffix == other.TrafficManagerDnsSuffix
-                       && BatchEndpointResourceId == other.BatchEndpointResourceId;
+                       && BatchEndpointResourceId == other.BatchEndpointResourceId
+                       && AzureOperationalInsightsEndpointResourceId == other.AzureOperationalInsightsEndpointResourceId; ;
```
Oops. :) Fixed.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/Commands.OperationalInsights.Test.csproj:
```diff
@@ -210,43 +215,7 @@
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="SessionRecords\Microsoft.Azure.Commands.OperationalInsights.Test.WorkspaceTests\TestWorkspaceEnableDisableListIntelligencePacks.json">
+    <None Include="SessionRecords\**\*">
```
Hm, interesting. VS now seems to prefer **\* in net core projects so I figured this was a shift we were making (and I'll admit I prefer this approach). But if this is a paradigm you all would like to follow, I'll respect that.

Fixed.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/ScenarioTests/QueryTests.cs:
```diff
@@ -0,0 +1,70 @@
+        static QueryTests()
+        {
+            //Uncomment this to regenerate the SessionRecords
+            //Environment.SetEnvironmentVariable("AZURE_TEST_MODE", "Record");
```
I found using the environment variables *super* frustrating and launching VS from a cmdline definitely wasn't fun. I've removed the comments but I'll probably continue using this approach to run the tests.

maddieclayton replied:
That's fine as long as they aren't pushed to preview.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/ScenarioTests/QueryTests.cs:
```diff
@@ -0,0 +1,70 @@
+
+        private EnvironmentSetupHelper helper;
+
+        public QueryTests()
```
The main reason this was done was that the existing base tests do additional setup that isn't needed for the data plane tests. However, I've refactored that base class to support the data plane tests.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/ScenarioTests/QueryTests.ps1:
```diff
@@ -0,0 +1,34 @@
+    $wsId = "DEMO_WORKSPACE"
+    $query = "union * | take 10"
+
+    $results = Invoke-AzureRmOperationalInsightsQuery -WorkspaceId $wsId -Query $query -UseDemoKey
```
Oops! I was copying some existing tests we have for sanity checking the API results and totally forgot to copy the asserts. :(

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/ScenarioTests/QueryTests.ps1:
```diff
@@ -0,0 +1,34 @@
+    $query = "union * | take 10"
+    $timespan = (New-Timespan -Hours 1)
+
+    $results = Invoke-AzureRmOperationalInsightsQuery -WorkspaceId $wsId -Query $query -Timespan $timespan -UseDemoKey
```
Same thing. Oops. :(

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+﻿using Microsoft.Azure.Commands.Common.Authentication;
```
I've added something to the OperationalInsights changelog. Is there any other changelog that needs to be updated? Including something for the changes I had to make to the auth project?

maddieclayton replied:
No, the addition to the OperationalInsights changelog should be fine.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+    public class InvokeOperationalInsightsQuery : ResourceManager.Common.AzureRMCmdlet
+    {
+        [Parameter(Mandatory = true, HelpMessage = "The workspace ID.")]
+        public string WorkspaceId { get; set; }
```
Fixed.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+    {
+        [Parameter(Mandatory = true, HelpMessage = "The workspace ID.")]
+        public string WorkspaceId { get; set; }
+
```
Fixed. This now supports piping in a PSWorkspace by extracting the CustomerId as the WorkspaceId. I know that sounds like a wrong mapping, but it is just a confusing nomenclature issue.

csuich2 commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            OperationalInsightsDataClient.WorkspaceId = WorkspaceId;
+
+            if (Wait != int.MinValue)
```
I generally forget about nullable...so thanks :)

csuich2 commented:
@maddieclayton Thanks for all the feedback! (and sorry I missed a bunch of stuff). I believe I've addressed everything with the exception of ShouldProcess, which I don't believe applies. However, I left a few follow up questions. I'll be sending the email for cmdlet review shortly.

markcowl commented:
@csuich2 Please respond to review feedback: https://github.com/Azure/azure-powershell-pr/wiki/Operational-Insights-Data#invoke-azurermoperationallinsightsquery

csuich2 commented:
@markcowl Hey Mark - thanks for the feedback! A couple clarifications:

1) ResourceId - The only issue I see here is due to an oddity in how our data plane works. The data plane can't actually access resources from the ARM ResourceId. It needs another piece of information that the user either supplies, or can be obtained from the full ARM object. Is there precedent for another data plane cmdlet having to invoke the management plane cmdlet in order to obtain this info?

2) AsJob - It certainly makes sense to implement AsJob, but the link you put there goes to the cmdlet review itself, not a page in the wiki. I tried searching the wiki for AsJob, but couldn't find anything. Can you fix that link? That said, the special cmdlet parameter that we are exposing actually allows the user to specify how long the data plane API should wait for the DB to return results, not how the timeout between the client and API. This is very useful for customers as our default time to wait for the DB is shorter than some customers need, so we need to provide this option for customers to specify a much longer time to wait.

3) Great point. So this is what it looks like by default:
``` PowerShell
$results = Invoke-AzureRmOperationalInsightsQuery -WorkspaceId "63613592-b6f7-4c3d-a390-22ba13102111" -Query "union * | take 2"
$results.Results


TenantId                 : 63613592-b6f7-4c3d-a390-22ba13102111
SourceSystem             : OpsManager
TimeGenerated            : 2017-12-13T14:59:59Z
Solution                 : LogManagement
...

TenantId                 : 63613592-b6f7-4c3d-a390-22ba13102111
SourceSystem             : OpsManager
TimeGenerated            : 2017-12-13T16:59:59Z
Solution                 : LogManagement
...
```

Printing $results.Results does show in table view, but it can be a bit hard to read since there are a LOT of rows. If you project to only a few rows, it does look much better.
``` PowerShell
TenantId                             SourceSystem TimeGenerated        Solution
--------                             ------------ -------------        --------
63613592-b6f7-4c3d-a390-22ba13102111 OpsManager   2017-12-13T11:59:59Z LogManagement
63613592-b6f7-4c3d-a390-22ba13102111 OpsManager   2017-12-13T12:59:59Z LogManagement
```

Any suggestions on how to make the first case better?

maddieclayton commented:
@azuresdkci Test this please

csuich2 commented:
@markcowl Looks like the CI build is failing, but with the below errors. I didn't think I broke those...

12:07:45 "d:\workspace\powershell\build.proj" (Full target) (1) ->
12:07:45 (DependencyAnalysis target) -> 
12:07:45   d:\workspace\powershell\build.proj(510,5): error MSB3073: The command "d:\workspace\powershell\src\Package\StaticAnalysis.exe d:\workspace\powershell\src\Package\Debug d:\workspace\powershell\src\Package true false" exited with code 255.
12:07:45 
12:07:45 
12:07:45 "d:\workspace\powershell\build.proj" (Full target) (1) ->
12:07:45 (InvokeXUnit target) -> 
12:07:45   d:\workspace\powershell\src\ResourceManager\Profile\Commands.Profile.Test\AzureRmProfileTests.cs(846): error : Microsoft.Azure.Commands.ResourceManager.Common.Test.AzureRMProfileTests.SavingProfileWorks [FAIL] [d:\workspace\powershell\build.proj]
12:07:45   d:\workspace\powershell\src\ResourceManager\Profile\Commands.Profile.Test\PSSerializationTests.cs(86): error : Microsoft.Azure.Commands.Profile.Test.PSSerializationTests.CanConvertProfieWithCustomEnvironment [FAIL] [d:\workspace\powershell\build.proj]
12:07:45 
12:07:45     523 Warning(s)
12:07:45     3 Error(s)

markcowl commented:
@csuich2 I'll take a look at this this afternoon and create a PR into your branch with the common cod echanges

markcowl commented:
@csuich2 Created a PR into your fork which should resolve the common code issues.

csuich2 commented:
@markcowl I tested the changes locally & my tests & manual verification seemed fine. However, the build is still failing due to static analysis.

https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/1525/console

csuich2 commented:
@azuresdkci Test this please

markcowl commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace Microsoft.Azure.Commands.OperationalInsights.Query
+{
+    [Cmdlet("Invoke", "AzureRmOperationalInsightsQuery"), OutputType(typeof(PSQueryResponse))]
```
Since you have multiple parameter sets, you must specify a DefaultParameterSetName

markcowl commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/packages.config:
```diff
@@ -1,4 +1,5 @@
 <packages>
   <package id="Microsoft.Azure.Management.OperationalInsights" version="0.18.0-preview" targetFramework="net45" />
-</packages>
+  <package id="Microsoft.Azure.OperationalInsights" version="0.9.0-preview" targetFramework="net452" />
```
Note - you will need to add this assembly to the psd1 file for your module in the 'RequiredAssemblies' property

markcowl commented on src/ResourceManager/OperationalInsights/AzureRM.OperationalInsights.psd1:
```diff
@@ -145,8 +146,7 @@ PrivateData = @{
         # ReleaseNotes of this module
-        ReleaseNotes = '* Add support for online help
-    - Run Get-Help with the -Online parameter to open the online help in your default Internet browser'
+        ReleaseNotes = 'https://github.com/Azure/azure-powershell/blob/dev/ChangeLog.md'
```
You don't need to update the version or the release notes.

You will also need to add the new dataplane sdk assembly into the 'requiredassemblies' proeprty in this psd1

markcowl commented on src/ResourceManager/OperationalInsights/ChangeLog.md:
```diff
@@ -21,6 +21,9 @@
 * Added Location Completer to -Location parameters allowing tab completion through valid Locations
 * Added ResourceGroup Completer to -ResourceGroup parameters allowing tab completion through resource groups in current subscription
 
+## Version 4.0.1
```
Do not add the extra heading, please put this after the comments above, with no heading in between - changelogs versiosn are updated as part of the release.

csuich2 commented:
@markcowl It looks like the build failure was a transient issue. Looks like everything else is passing and I addressed all the feedback from your last set of comments. I also saw that you updated the link for the -AsJob option, so I have updated the cmdlet to support that as well as regenerated the docs.

markcowl commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/help/Invoke-AzureRmOperationalInsightsQuery.md:
````diff
@@ -78,6 +78,19 @@ See [https://dev.loganalytics.io/documentation/Using-the-API/RequestOptions](htt
 ## PARAMETERS
 
+### -AsJob
+Run cmdlet in the background```yaml
````
The ```yaml comment needs to be on its own line - unfortunately, platyPS sometimes does this with added properties

alexeldeib replied:
Fixed!

alexeldeib commented on src/ResourceManager/Profile/ChangeLog.md:
```diff
@@ -22,6 +22,8 @@
 * Add-AzureRmAccount
   * Added -MSI login for authenticationg using the credentials of the Managed Service Identity of the current VM / Service
   * Fixed KeyVault Authentication when logging in with user-provided access tokens
+* Add-AzureRmEnvironment and Set-AzureRmEnvironment
```
Added

alexeldeib commented:
Thanks, Mark. I’ve made those changes, and I’ll keep an eye on the build as it progresses. Barring further issues, appreciate your work getting us to to the finale stage here!

markcowl commented:
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/387/

cormacpayne commented on src/ResourceManager/OperationalInsights/ChangeLog.md:
```diff
@@ -23,6 +23,7 @@
 * Added Location Completer to -Location parameters allowing tab completion through valid Locations
 * Added ResourceGroup Completer to -ResourceGroup parameters allowing tab completion through resource groups in current subscription
 * Obsoleted -Tags in favor of -Tag for New-AzureRmOperationalInsightsSavedSearch, Set-AzureRmOperationalInsightsSavedSearch, New-AzureRmOperationalInsightsWorkspace, and Set-AzureRmOperationalInsightsWorkspace
+* Added support for V2 API querying via Invoke-AzureRmOperationalInsightsQuery. See [https://dev.loganalytics.io/](https://dev.loganalytics.io/) for more info on the new API.
```
@csuich2 @alexeldeib please move this line you added to be under the `## Current Release` header -- merging in the latest changes from the `preview` branch moved this under the latest release's notes

cormacpayne commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights.Test/ScenarioTests/QueryTests.cs:
```diff
@@ -0,0 +1,37 @@
+﻿using Xunit;
```
@csuich2 @alexeldeib it looks like a few `.cs` files that were added are missing a license header -- please add one using the header from any existing file

cormacpayne commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Models/PSQueryResponse.cs:
```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.Azure.OperationalInsights.Models;
```
@csuich2 @alexeldeib license header

cormacpayne commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,111 @@
+﻿using Microsoft.Azure.Commands.Common.Authentication;
```
@csuich2 @alexeldeib license header

cormacpayne commented on src/ResourceManager/OperationalInsights/Commands.OperationalInsights/Query/InvokeOperationalInsightsQuery.cs:
```diff
@@ -0,0 +1,111 @@
+
+namespace Microsoft.Azure.Commands.OperationalInsights.Query
+{
+    [Cmdlet("Invoke", "AzureRmOperationalInsightsQuery", DefaultParameterSetName = ParamSetNameByWorkspaceId),
```
@csuich2 @alexeldeib nit: I would change `"Invoke"` to `VerbsLifecycle.Invoke`

cormacpayne commented on src/ResourceManager/Profile/ChangeLog.md:
```diff
@@ -27,6 +27,8 @@
 * Add-AzureRmAccount
   * Added -MSI login for authenticationg using the credentials of the Managed Service Identity of the current VM / Service
   * Fixed KeyVault Authentication when logging in with user-provided access tokens
+* Add-AzureRmEnvironment and Set-AzureRmEnvironment
```
@csuich2 @alexeldeib same comment about moving the lines you added to be under the `## Current Release` header

cormacpayne commented on src/ResourceManager/Profile/Commands.Profile/Environment/AddAzureRMEnvironment.cs:
```diff
@@ -151,6 +151,14 @@ public string DataLakeAudience
 
+        [Parameter(Position = 21, Mandatory = false, ValueFromPipelineByPropertyName = true,
+           HelpMessage = "The audience for tokens authenticating with the Azure Log Analytics API.")]
+        public string AzureOperationalInsightsEndpointResourceId { get; set; }
```
@csuich2 @alexeldeib please update the markdown files for `Add-AzureRmEnvironment` and `Set-AzureRmEnvironment` to reflect the new parameters in the cmdlets

alexeldeib commented:
Thanks @cormacpayne, all addressed! I also added BatchEndpointResourceId to the help file for the Set/Add-AzureEnvironment commands, which I noticed was in the C# but missing from the markdown.

alexeldeib commented:
@cormacpayne I noticed a transient build failure, but seems to have resolved on a fresh run. Let me know if there are additional changes to be made on my side.

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/405/
